swiftmt/parser.bal

```diff
@@ -54,6 +54,81 @@ public isolated function parseSwiftMt(string finMessage) returns record {}|error
     return xmldata:parseAsType(customizedXml, {textFieldName: "content"}, t = messageRecord);
 }
 
+# Converts a SWIFT message record into a FIN message string.
+#
+# This function processes a SWIFT message record and maps it to the appropriate SWIFT FIN message type based on
+# the message type and validation flag. It converts the record into a FIN message using the Prowide Conversion Service.
+# The function supports multiple validation flags such as `STP`, `REMIT`, and `COV`, and also handles messages with 
+# a type greater than or equal to 90.
+#
+# + message - The SWIFT message record to be converted.
+# + return - Returns the converted FIN message as a `string`, or an error if the message type is invalid or unsupported.
+public isolated function getFinMessage(record {} message) returns string|error {
+    xml swiftMessageXml = check xmldata:toXml(message, options = {textFieldName: "content"});
+    string messageType = (swiftMessageXml/**/<messageType>).data();
+    string validationFlag = (swiftMessageXml/**/<ValidationFlag>/<value>).data();
+    typedesc<record {}>? messageRecord;
+    if validationFlag.equalsIgnoreCaseAscii("STP") || validationFlag.equalsIgnoreCaseAscii("REMIT") || validationFlag.equalsIgnoreCaseAscii("COV") {
+        messageRecord = messageMapper[messageType + validationFlag];
+        if messageRecord is () {
+            return error("SWIFT message type is invalid or not supported.");
+        }
+    }
+    if check int:fromString(messageType.substring(1, 3)) >= 90 {
+        messageRecord = messageMapper[messageType.substring(1, 3)];
+        if messageRecord is () {
+            return error("SWIFT message type is invalid or not supported.");
+        }
+    }
+    messageRecord = messageMapper[messageType];
+    if messageRecord is () {
+        return error("SWIFT message type is invalid or not supported.");
+    }
+    prowide:ConversionService srv = prowide:newConversionService1();
+    return srv.getFIN(convertToProprietaryXml(swiftMessageXml).toString());
+}
+
+# Converts a SWIFT message XML into a proprietary XML format for FIN message generation.
+#
+# This function takes a SWIFT message in XML format and transforms it into a proprietary format required
+# for generating FIN messages. 
+#
+# + swiftMessageXml - The SWIFT message XML to be converted.
+# + return - Returns the transformed proprietary XML.
+isolated function convertToProprietaryXml(xml swiftMessageXml) returns xml{
+    xml block4Xml = xml ``;
+    xml finalBlockXml = xml ``;
+    xml:Element finalSwiftXml = xml `<message/>`;
+    foreach xml:Element tagElement in (swiftMessageXml/**/<block4>).elementChildren() {
+        if tagElement.getName().equalsIgnoreCaseAscii("Transaction")  || 
+           tagElement.getName().equalsIgnoreCaseAscii("UndrlygCstmrCdtTrf") {
```

Comment:
Shall we add a comment explaining this? Why only these 2 elements are considered for block 4

Reply:
fixed in `76432da`